src/clj/disaster/middleware/oauth2_jwt.clj

```diff
@@ -0,0 +1,197 @@
+;; clojure map containing a :public-key and optionally a :private-key.
+(defonce keystore
+  (atom {}))
+
+
+(defn- resolve-key
+  "Returns java.security.Key given key-fn, jwks-url and :key-type in jwt-header.
+  If no key is found refreshes"
+  [key-type jwks-url jwt-header]
+  (log/debug "Resolving key" jwt-header "from jwk cache for" jwks-url)
+  (let [key-fn (fn [] (get-in @keystore [jwks-url (:kid jwt-header) key-type]))]
+    (if-let [key (key-fn)]
+      key
+      (do (log/debug "Fetch and resolve key" jwt-header "from" jwks-url)
+          (when-let [new-keys (fetch-keys jwks-url)]
+            (swap! keystore #(update % jwks-url merge new-keys)))
+          (if-let [key (key-fn)]
+            key
+            (do
+              (log/error "Could not locate public key corresponding to jwt header's kid:" (:kid jwt-header) "for url:" jwks-url)
+              (throw (ex-info (str "Could not locate key corresponding to jwt header's kid: " (:kid jwt-header) " for url: " jwks-url)
+                              {:type :validation :cause :unknown-key}))))))))
+
+
+(s/fdef resolve-public-key
+  :args (s/cat :jwks-url    ::jwks-url
+               :jwt-header  ::jwt-header)
+  :ret  ::public-key)
+
+(def resolve-public-key
+  "Returns java.security.PublicKey given jwks-url and :kid in jwt-header.
+  If no key is found refreshes"
+  (partial resolve-key :public-key))
+
+
+(s/fdef resolve-private-key
+  :args (s/cat :jwks-url ::jwks-url
+               :jwt-header ::jwt-header)
+  :ret  ::private-key)
+
+(def resolve-private-key
+  (partial resolve-key :private-key))
+
+
+(s/fdef unsign
+  :args (s/cat :jwks-url ::jwks-url
+               :token    ::jwt)
+  :ret  ::claims)
+
+(defn- remove-bearer [token]
+  (if (and token (str/starts-with? (str/lower-case token) "bearer "))
+    (subs token (count "Bearer "))
+    token))
+
+(defn unsign
+  "Given jwks-url, token, and optionally opts validates and returns the claims
+  of the given json web token. Opts are the same as buddy-sign.jwt/unsign."
+  ([jwks-url token]
+   (unsign jwks-url token {}))
+  ([jwks-url token opts]
+   (assert (s/valid? ::jwks-url jwks-url) "jwks-url must conform to ::jwks-url")
+   (let [token (remove-bearer token)]
+     (assert (s/valid? ::jwt token) "token must conform to ::jwt")
+     (jwt/unsign token (partial resolve-public-key jwks-url) (merge {:alg :rs256} opts)))))
+
+(defn scopes
+  "Given the claims from unsign returns the jwt scope as a set of strings.
+
+  For a jwt without scope, an empty set will be returned."
+  [claims]
+  (assert (map? claims) "claims must be a map!")
+  (if-let [claims (not-empty (get claims :scope))]
+    (do
+      (assert (string? claims) ":scope in claims must be a string!")
+      (into (sorted-set) (str/split claims #"\s+")))
+    #{}))
+
+(s/fdef sign
+  :args (s/cat :jwks-url ::jwks-url
+               :kid      ::kid
+               :claims   ::claims)
+  :ret ::jwt)
+
+(defn sign
```

Comment:
spec this boi with multi-arity spec as well. https://blog.taylorwood.io/2017/10/15/fspec.html